ksqldb-engine/src/main/java/io/confluent/ksql/engine/TransientQueryCleanupService.java

```diff
@@ -0,0 +1,312 @@
+  private final Pattern transientPattern;
+  private static final Pattern LEAKED_TOPIC_PREFIX_PATTERN =
+          Pattern.compile("(?i).*transient_.*_[0-9]\\d*_[0-9]\\d*");
+
+  private final BlockingQueue<Callable<Boolean>> stateDirsCleanupTasks;
+  private final BlockingQueue<Callable<Boolean>> topicsCleanupTasks;
+  private Set<String> localCommandsQueryAppIds;
+  private boolean isLocalCommandsInitialized;
+  private boolean isLocalCommandsProcessed;
+  private QueryRegistry queryRegistry;
+  private final String stateDir;
+  private final ServiceContext serviceContext;
+  private final int initialDelay;
+  private final int intervalPeriod;
+
+
+  public TransientQueryCleanupService(final ServiceContext serviceContext,
+                                      final KsqlConfig ksqlConfig) {
+    this.transientPattern = Pattern.compile(
+            ksqlConfig.getString(
+                    KsqlConfig.KSQL_TRANSIENT_QUERY_NAME_PREFIX_CONFIG));
+
+    this.initialDelay = ksqlConfig.getInt(
+            KsqlConfig.KSQL_TRANSIENT_QUERY_CLEANUP_SERVICE_INITIAL_DELAY_SECONDS);
+
+    this.intervalPeriod = ksqlConfig.getInt(
+            KsqlConfig.KSQL_TRANSIENT_QUERY_CLEANUP_SERVICE_PERIOD_SECONDS);
+
+    topicsCleanupTasks = new LinkedBlockingDeque<>();
+    stateDirsCleanupTasks = new LinkedBlockingDeque<>();
+
+    this.stateDir = ksqlConfig.getKsqlStreamConfigProps()
+            .getOrDefault(
+                    StreamsConfig.STATE_DIR_CONFIG,
+                    StreamsConfig.configDef()
+                            .defaultValues()
+                            .get(StreamsConfig.STATE_DIR_CONFIG))
+            .toString();
+
+    this.serviceContext = serviceContext;
+    this.isLocalCommandsInitialized = false;
+    this.isLocalCommandsProcessed = false;
+  }
+
+  @Override
+  protected void runOneIteration() {
+    if (isLocalCommandsInitialized && !isLocalCommandsProcessed) {
+      LOG.info("Adding LocalCommands to TransientQueryCleanupService.");
+      localCommandsQueryAppIds.forEach(id -> {
+        addTopicCleanupTask(new TransientQueryTopicCleanupTask(serviceContext, id));
+        addStateCleanupTask(new TransientQueryStateCleanupTask(id, stateDir));
+      });
+      isLocalCommandsProcessed = true;
+    }
+
+    LOG.info("Starting cleanup for leaked resources.");
+
+    topicsCleanupTasks.addAll(findPossiblyLeakedTransientTopics());
+    stateDirsCleanupTasks.addAll(findPossiblyLeakedStateDirs());
+
+    final int numLeakedTopics = topicsCleanupTasks.size();
+    final int numLeakedStateDirs = stateDirsCleanupTasks.size();
+
+    LOG.info("Cleaning up {} leaked topics.", numLeakedTopics);
+    cleanupLeakedTopics();
+
+    LOG.info("Cleaning up {} leaked state directories.", numLeakedStateDirs);
+    cleanupLeakedStateDirs();
+  }
+
+  @Override
+  public Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(initialDelay, intervalPeriod, TimeUnit.SECONDS);
```

Comment:
Sorry, one more quick question about this: What's the behavior if an unhandled exception kills this thread? Does the scheduler just create a new one the next time around? In that case, do we have any concerns about tasks left behind in our task lists from the prior, incomplete run?

Reply:
good catch. Any exception escaping the run method halts all further work, without notice. from the docs: ```Run one iteration of the scheduled task. If any invocation of this method throws an exception, the service will transition to the Service.State.FAILED state and this method will no longer be called.```
added a try-catch block and logging the failure to safeguard against that.